Reports: include read_only tag categories

cben opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1388659
## Purpose or Intent

Split off [discussion on #9914](https://github.com/ManageIQ/manageiq/pull/9914#issuecomment-234917815).
Reports editor hides any `read_only` tag categories.
I suspect the condition was simply copied (it's exactly same line) from 
https://github.com/ManageIQ/manageiq/blob/62874a37/app/controllers/application_controller/tags.rb#L265
https://github.com/ManageIQ/manageiq/blob/62874a37/app/controllers/application_controller/tags.rb#L316
but that code deals with user assigning/unassigning the tags, so forbidding read only makes sense;
reports merely want to _read_ tags which I think is fine.
## Effects
- Reports editor, first "Columns" tab:
  ![screenshot-localhost 3000 2016-08-14 16-51-03 available fields](https://cloud.githubusercontent.com/assets/273688/17649512/c01bda2e-623f-11e6-98eb-0ab52bde03dc.png)
- selecting the category in "Columns", allows manipulating it in other tabs ("Consolidation", "Styling" etc) 
- Reports editor, "Filters" tab, with "Tag" expression:
  ![screenshot-localhost 3000 2016-08-14 16-29-31 read_only cat](https://cloud.githubusercontent.com/assets/273688/17649437/26c66ffc-623e-11e6-8238-16ccc6c57822.png)
  ![screenshot-localhost 3000 2016-08-14 16-30-42 read_only ent](https://cloud.githubusercontent.com/assets/273688/17649440/2ede2324-623e-11e6-8d0b-9041885ad019.png)
- Chargeback assignments, "Tagged ...", categories dropdown:
  ![screenshot](https://gist.githubusercontent.com/cben/89e0232e2e5c7429637313651775968e/raw/955f85847514d8f74c0e14310f811bbdbbbad11c/ManageIQChargebackAssign.png)

[Where do read_only tags come from?  I don't know generally, my use case is tags auto-assigned from kubernetes labels (#7460, to be exposed in UI by #11591) which I want read_only so users don't assign them manually — refresh would overwrite that anyway.]
## Steps for Testing/QA
1. If #11591 and #11806 landed, use auto-tagging UI to create a tag:
   - Label something in OpenShift, e.g. `oc label node mynode foo=entry1`.
   - Add mapping from `foo` label key to `read_only cat 1` category.
   - Refresh the OpenShift provider.  Verify the labeled entity got a "read_only cat 1 : entry1" tag.
   
   (It's important that at least one tag actually got created — reports editor won't list tag categories that have no entries, even after this change.)
   
   Alternatively, create a read_only tag directly in `rails console`:
   
   ``` ruby
   cat = Classification.create_category!(name: "rocat1", description: "read_only cat 1", read_only: true)
   cat.add_entry(name: "roent1", description: "read_only entry 1")
   ```
2. In Cloud Intel -> Reports -> Create new (or Edit), verify "read_only cat 1" tag category is in list of columns.
3. In Cloud Intel -> Chargeback -> Assignments -> Compute, choose "Assign To: Tagged VMs and Instances" and verify "read_only cat 1" is in "Tag Category" dropdown.
4. In Cloud Intel -> Chargeback -> Assignments -> Storage, choose "Assign To: Tagged Datastores" and verify "read_only cat 1" is in "Tag Category" dropdown.

@miq-bot add-labels reporting, ui, bug, darga/yes

cben commented:
There is another such line in
https://github.com/ManageIQ/manageiq/blob/62874a37/app/controllers/application_controller/performance.rb#L1481
which might also need fixing(?), I'm not familiar with that.

cben commented:
@zeari Please review.
How should I test this?  I now only checked UI works.

zeari commented:
> How should I test this? I now only checked UI works.

You can tag an entity manually and see that it shows up.

@gtanzillo You once said that the reason for omitting read only tags was that there were internal tags that the user wasnt supposed to see, is this change still ok?

cben commented:
OK, here is a problem: they get added under "My Company Tags".
[This doc][1] explains "Company Tags" are ones you define and assign, vs "System Tags" which are automatically assigned.  Makes sense, and *sounds* like read_only tags should be listed as "System Tags".
@gtanzillo I can't find any code dealing with "System Tags".  Do you know if those are actually a thing?  Is it a concept we want to stick to?

[1]: https://access.redhat.com/documentation/en/red-hat-cloudforms/4.1/general-configuration/#company_tag_categories_and_tags

simon3z commented:
@gtanzillo I think there's an open question to you here. The rest LGTM :+1: 

@moolitayer @enoodle can you review?
@miq-bot assign cben

moolitayer commented:
LGTM per @gtanzillo's comment [here](https://github.com/ManageIQ/manageiq/pull/9914#issuecomment-234959299)

simon3z commented:
LGTM :+1: 

@miq-bot assign chessbyte

jrafanie commented:
@cben I'm not familiar with this code but I tracked down this 2007!  The author has been suppressed to protect the innocent.  It seems like this was done purposely ON assignment only as you guessed.  If you can confirm after reviewing this commit, I think we can merge.  Please ping me if you're good with it.

```
commit 590d6b35e5df6a8cf9dae2f5f1eae66757ededf3
Date:   Wed Oct 31 00:27:16 2007 +0000

    BugzId:
    Fixed bug where "read only" tags were displayed in the My Company Tags assignment screen

    git-svn-id: http://miq-ubuntusub.manageiq.com/svn/svnrepos/Manageiq/trunk@4504 9653ab01-789c-4043-9faa-2b6746af0d71

diff --git a/vmdb/app/controllers/application.rb b/vmdb/app/controllers/application.rb
index 269cab4..87880da 100644
--- a/vmdb/app/controllers/application.rb
+++ b/vmdb/app/controllers/application.rb
@@ -670,16 +670,15 @@ class ApplicationController < ActionController::Base

        # Build the classification assignment screen
        def classify_build_screen
-               cats = Classification.categories        .sort{|a,b| a.name <=> b.name}  # Get the categories, sort by name
+               cats = Classification.categories.sort{|a,b| a.name <=> b.name}  # Get the categories, sort by name
                @categories = Hash.new          # Classifications array for first chooser
+               cats.delete_if{ |c| c.readonly != "0" || c.entries.length == 0} # Remove categories that are read only or have no entries
                cats.each do |c|
-                       if c.readonly == "0"            # Only show the non-read only categories
                                if c.single_value == "1"
                                        @categories[c.description + " *"] = c.name
                                else
                                        @categories[c.description] = c.name
                                end
-                       end
                end
                session[:cat] ||= cats.first                                                                                                                                            # Set to first category, if not already set

@@ -687,10 +686,13 @@ class ApplicationController < ActionController::Base

                session[:assignments] = Classification.find_assigned_entries(@tagitems[0])              # Start with the first items assignments
                @tagitems.each do |item|
-                       itemassign = Classification.find_assigned_entries(item)                                                                                         # Get each items assignments
-                       session[:assignments].delete_if { |a|   !itemassign.include?(a) }                                                       # Remove any assignments that are not in the new items assignments
-                       break if session[:assignments].length == 0                                                                                                                                                      # Stop looking if no assignements are left
+                       itemassign = Classification.find_assigned_entries(item)                                                 # Get each items assignments
+                       session[:assignments].delete_if { |a|   !itemassign.include?(a) }       # Remove any assignments that are not in the new items assignments
+                       break if session[:assignments].length == 0                                                                                                      # Stop looking if no assignments are left
                end
+               if session[:assignments].length > 0                                                                                                                                                                                     # if any assignments left
+                       session[:assignments].delete_if { |a| a.parent.readonly != "0"}         # remove the ones from read only categories
+    end
                classify_build_entries_pulldown
        end

```

jrafanie commented:
Also, @cben please see euwe label if desired.

cben commented:
@jrafanie Thanks for doing the archeology.  Makes sense, I'm good with it.

@miq-bot add-label euwe/yes
@simon3z This is necessary for reporting based on auto-tagging.  Do we want it even in darga?

simon3z commented:
> @simon3z This is necessary for reporting based on auto-tagging. Do we want it even in darga? EDIT: I see we originally wanted this darga/yes — do we still?

@cben did we have a BZ with anyone requesting this in 4.1?

cben commented:
We don't have any BZ for this change itself.  @zeari We have various BZs for reporting by tags, and my understanding that some/all of them we also want to work with auto-tagged read only tags — will/did any of these ship in a darga-* release?
E.g. I see #10671 has been darga/backported.

P.S. @jrafanie can you say anything about [this remaining occurence][1]?  Its result gets used [in `_perf_options` layout][2], for "EmsCluster", "Host", "Storage", "AvailabilityZone" models.  I have no idea about any of that...
[1]: https://github.com/ManageIQ/manageiq/blob/62874a37/app/controllers/application_controller/performance.rb#L1481 
[2]: https://github.com/ManageIQ/manageiq/search?utf8=%E2%9C%93&q=perf_options%5B%3Acats%5D

zeari commented:
> will/did any of these ship in a darga-* release?

@cben this one too: https://github.com/ManageIQ/manageiq/pull/10940

A new BZ is probably the way to go

jrafanie commented:
> We don't have any BZ for this change itself. @zeari We have various BZs for reporting by tags, and my understanding that some/all of them we also want to work with auto-tagged read only tags — will/did any of these ship in a darga-* release?
> E.g. I see #10671 has been darga/backported.


@cben this commit doesn't have a BZ associated with it.  If we want to ship this fix, please add a BZ in the commit message.
 
> P.S. @jrafanie can you say anything about this remaining occurence? Its result gets used in _perf_options layout, for "EmsCluster", "Host", "Storage", "AvailabilityZone" models. I have no idea about any of that...

@cben I would think this was copy/pasted so if the context is building a list of tags for "view" purposes, then yes, it would be the same as this PR, we shouldn't filter them out.  If it's for assigning or modifying tags, then, I believe the check should remain.  Honestly, this logic should left out of the controllers and be a simple scope in the model:   "viewable_tags" or "assignable_tags" or something.  Either way, it's nasty that this conditional is litter everywhere.  For this issue, just fixing them is fine for now but scoping with a good scope name could really help the readability of this code.

@cben let me know when this is ready for merge (BZ in commit message), thanks!

zeari commented:
@jrafanie The BZ for this is https://bugzilla.redhat.com/show_bug.cgi?id=1382720
Please merge, we want this in euwe.

cc @simon3z

jrafanie commented:
@cben @zeari this one is really close, just missing a test or an explanation why there's no test and clarification on the commit message (why and what bz)

jrafanie commented:
@cben ping

cben commented:
Thanks for nudging (vacation + was busy with more urgent stuff).
Opened BZ, edited message.
Found another place in chargeback, but still haven't tested & screenshotted it.

Still need to add tests.
The only way I see to test is extract into small method, and test it doesn return a read_only cat.  Does that sound useful?

classifications.rb already has most of the scopes needed to make such queries pretty, but they're not in darga, so I'll do a followup PR.

@miq-bot add-label wip

simon3z commented:
@zeari please review

cben commented:
Added tests in the report editor.
Also exctracted (with tests) 2 identical codes for entries.

Added screenshot for the chargeback assignments fix.  (Got lazy and didnt add the nearly identical tests there; I don't want to maintain many nearly identical copies with tests, I'll try centralize all such code in Classification in a followup refactoring...)

@jrafanie @gtanzillo This is ready from my side.
@miq-bot remove-label wip

jrafanie commented on app/controllers/report_controller/reports/editor.rb:
```diff
@@ -1552,6 +1546,18 @@ def set_form_vars
 
+  def categories_hash
+    cats = Classification.categories.select { |c| c.show }
+    cats.delete_if { |c| c.entries.length == 0 }  # Omit categories for which entries dropdown would be empty.
```
suggest combining these two lines:
`cats = Classification.categories.select { |c| c.show && c.entries.size > 0 }`

Maybe in a followup PR, if this is a domain concept (visible categories with entries) we can move this logic to a scope in the model since we have the exact same logic in `get_categories_all` in the chargeback_controller.

I also see something that might be related:
```
app/controllers/miq_policy_controller/alert_profiles.rb:      c if !c.read_only? && c.show && c.entries.size > 0
lib/report_formatter/text.rb:      categories = Classification.categories.collect { |c| c if c.show }.compact
```

The first is visible categories with entries and the second is just visible categories.  It might be worth making these actual things in the model.

jrafanie commented on app/controllers/report_controller/reports/editor.rb:
```diff
@@ -1552,6 +1546,18 @@ def set_form_vars
+  end
+
+  def entries_hash(category_name)
+    cat = Classification.find_by_name(category_name)
```
prefer using `find_by` instead of dynamic finders.

`cat = Classification.find_by(:name => category_name)`

cben replied:
That's not a dynamic finder, it's [a class method](https://github.com/ManageIQ/manageiq/blob/5aa5c4adddc39d2762ea450b6d0ea6da9a8010fe/app/models/classification.rb#L330) that does a bit more, in particular it defaults to parent_id: 0 constraint which we need for a category.
Yeah the name is a bit confusing.  Do you think a PR renaming it to something like `Classification.find_category` would be a good idea?

jrafanie replied:
Thanks @cben.  Maybe it's worth seeing how hard it is to do.  There are an unknown number of callers of this method so it would depend how much work it is to do and verify.

jrafanie commented on spec/controllers/miq_report_controller/reports/editor_spec.rb:
```diff
@@ -159,6 +159,39 @@
+      it "returns non-empty categories" do
+        FactoryGirl.create(:classification_cost_center_with_tags)
+        FactoryGirl.create(:classification_department)
+        expect(controller.send(:categories_hash)).to eq("cc" => "Cost Center")
```
Nice, thanks for adding tests.

Where does "cc" and "Cost Center" come from and how do I know which factory created it?  How do I know which factory should be in the `categories_hash`?

This test is hard to understand because one has to open the factories to realize the first one is creating a classification named "cc" with children, whereas the second one just creates a classification with no children.  Honestly, I think it would be beneficial to use the base classification factory and customize the parts that are important to this test.

jrafanie commented on spec/controllers/miq_report_controller/reports/editor_spec.rb:
```diff
@@ -159,6 +159,39 @@
+      end
+
+      it "omits show: false categories" do
+        FactoryGirl.create(:classification_cost_center_with_tags, :show => false)
```
Honestly, a toplevel `let(:classification) { FactoryGirl.create(...) }` that you then do `classification.update(:show => false) here would be helpful.

jrafanie commented on spec/controllers/miq_report_controller/reports/editor_spec.rb:
```diff
@@ -159,6 +159,39 @@
+
+    describe "#entries_hash" do
+      it "returns entries" do
+        FactoryGirl.create(:classification_department_with_tags)
```
Same ^

cben commented:
Done, PTAL.
Used methods instead of `let` because I dislike relying on let's on-demand execution for side effect.

simon3z commented:
> Done, PTAL.
> Used methods instead of let because I dislike relying on let's on-demand execution for side effect.

@cben @jrafanie is this mergeable today? Can we make an effort of getting this reviewed/revised/merged/backported as well so we'll deliver the full auto-tagging capability today?

zeari commented:
LGTM @simon3z @jrafanie

jrafanie commented:
> @cben @jrafanie is this mergeable today? Can we make an effort of getting this reviewed/revised/merged/backported as well so we'll deliver the full auto-tagging capability today?

@simon3z It would be nice to make an effort to fix the simple rubocops. The bot is there so you can get feedback faster than a human reviewer.  I'd rather not have to merge this now and have a followup PR that also needs backport to fix whitespace.  Let's get it right the first time to avoid git conflicts.

@cben looks great other than the whitespace complaints.

chessbyte commented:
@cben Let's get the Rubocops addressed as soon as possible.  /cc @simon3z

cben commented:
Can someone help me understand what Rubocop wants in this instance?
It doesn't reproduce when I run `rubocop` locally (prob. using wrong config) and the block braces and |arg| look consistent with all the file and our style guide.

Fryguy commented:
Line 1557 and 1558 have trailing whitespace

jrafanie commented:
@cben Thanks for looking into the cops.  I'm not exactly sure what it's complaining about.  I'll give others a few minutes to review to see if they understand why the bot is concerned about those lines.  Then, I'll merge.  Next time, if you see bot comments that seem totally wrong, please speak up so we can keep them useful and not noisy with useless messages.

cben commented:
Changed `.size > 0` to `!...empty?` (which bot didn't mention but local rubocop legitimately suggested).
Still at a loss about whitespace.  Thanks @Fryguy but I do not see any trailing whitespace (with eyes nor `grep`).  Adding such whitespace on a test file leads to different message – `Trailing whitespace detected.` so it's not that.

Fryguy commented:
Oh right...I forgot there was a different message for that.

Yeah, I don't know what's going on with the bot...we'll have to take a look at it.

jrafanie commented:
> Checked commits cben@71b9736~...0d9d73c with ruby 2.2.5, rubocop 0.37.2, and haml-lint 0.16.1
> 3 files checked, 0 offenses detected
> Everything looks good. ⭐️

What now bot!!!  👏 🚢 :shipit:

chrisarcand commented on spec/controllers/miq_report_controller/reports/editor_spec.rb:
```diff
@@ -174,6 +174,47 @@
+      end
+
+      it "omits show: false categories" do
+        non_empty_category.update(:show => false)
```
> Used methods instead of let because I dislike relying on let's on-demand execution for side effect

Memoized helpers; the reason you would want to use them here is that now you are writing tests like this that require multiple database writes and slowing your tests down. You should either inline what you need in this example, wrap it in a context block and use `before` with an instance var, or (in my opinion) use a memoized helper to change the attributes given to this object on create. There are many ways to avoid what you're doing here.

cben replied:
Do you that memoized `let` would be better than `def`?  IIUC the transaction is rolled back between every example, so if I use `let` in two examples it would still create it twice.

I think you mean `.update` is doing a 2nd unnecessary write.  Fixing...

chrisarcand replied:
> so if I use let in two examples it would still create it twice.

No, the first would not be fired as it was overridden.

cben replied:
Made the overrides params of the helper passed into the factory, writing once *per example that calls it*.  PTAL.
I suppose I could have `let` that doesn't save and/or shadow the `let` but that sounds complicated, if you still don't like it I'd rather inline.

I don't see a simple way to not write at all (.build / .stub), because the funcs under test query the whole table with scopes etc.

chrisarcand replied:
I don't know what you mean here. You can change the parameters of a single let within another context to only have one write, and it looks pretty spiffy. I can show you an example sometime if you like. 

> Made the overrides params of the helper passed into the factory, writing once per example that calls it. PTAL.

This is fine, just go with this.

chrisarcand commented on spec/controllers/miq_report_controller/reports/editor_spec.rb:
```diff
@@ -174,6 +174,47 @@
     end
+
+    def non_empty_category(attrs={})
+      cat = FactoryGirl.create(:classification, :name => "non_empty", :description => "Has entries", **attrs)
```
This at least removes the double write; that's fine 👍

chessbyte commented:
Euwe Backport details:
```sh
$ git log -1
commit 93ac929123683ea4bcf2d6e001a19ae5c2d1d667
Author: Joe Rafaniello <jrafanie@users.noreply.github.com>
Date:   Tue Nov 1 13:16:29 2016 -0400

    Merge pull request #10459 from cben/chargeback-read_only-tags
    
    Reports: include read_only tag categories
    (cherry picked from commit 81b121f00df79afc42c6730ccc16374133473edc)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1388659
```

chessbyte commented:
@simaishi Please make Euwe BZ for https://bugzilla.redhat.com/show_bug.cgi?id=1388659

chessbyte commented:
Darga Backport details:
```sh
$ git log -1
commit 5ac96cfa6432875b71a2235c4c09f976fd4b4c88
Author: Joe Rafaniello <jrafanie@users.noreply.github.com>
Date:   Tue Nov 1 13:16:29 2016 -0400

    Merge pull request #10459 from cben/chargeback-read_only-tags
    
    Reports: include read_only tag categories
    (cherry picked from commit 81b121f00df79afc42c6730ccc16374133473edc)
```

chessbyte commented:
@simaishi Please make Darga BZ for https://bugzilla.redhat.com/show_bug.cgi?id=1388659

simaishi commented:
Darga BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1391980
